misk-hibernate/src/main/kotlin/misk/hibernate/actions/HibernateDatabaseQueryMetadataFactory.kt

```diff
@@ -0,0 +1,320 @@
+package misk.hibernate.actions
+
+import misk.hibernate.Constraint
+import misk.hibernate.DbEntity
+import misk.hibernate.Id
+import misk.hibernate.Operator
+import misk.hibernate.Order
+import misk.hibernate.Projection
+import misk.hibernate.Property
+import misk.hibernate.Query
+import misk.hibernate.Select
+import misk.hibernate.Session
+import misk.hibernate.actions.HibernateDatabaseQueryDynamicAction.Companion.HIBERNATE_QUERY_DYNAMIC_WEBACTION_PATH
+import misk.hibernate.actions.HibernateDatabaseQueryStaticAction.Companion.HIBERNATE_QUERY_STATIC_WEBACTION_PATH
+import misk.security.authz.AccessAnnotationEntry
+import misk.web.MiskWebFormBuilder
+import misk.web.MiskWebFormBuilder.Field
+import misk.web.MiskWebFormBuilder.Type
+import misk.web.MiskWebFormBuilder.Companion.createEnumField
+import misk.web.MiskWebFormBuilder.Companion.createSyntheticEnumField
+import misk.web.MiskWebFormBuilder.Companion.maybeCreatePrimitiveField
+import misk.web.metadata.DatabaseQueryMetadata
+import javax.inject.Inject
+import javax.inject.Singleton
+import javax.persistence.Table
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.KType
+import kotlin.reflect.full.allSupertypes
+import kotlin.reflect.full.createType
+import kotlin.reflect.full.declaredMemberFunctions
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.memberProperties
+
+@Singleton
+class HibernateDatabaseQueryMetadataFactory @Inject constructor(
+  val accessAnnotationEntries: List<AccessAnnotationEntry>
+) {
+  fun <T : DbEntity<T>> fromQuery(
+    dbEntityClass: KClass<T>,
+    queryClass: KClass<out Query<T>>? = null,
+    accessAnnotationClass: KClass<out Annotation>,
+  ): DatabaseQueryMetadata {
+    val table = dbEntityClass.findAnnotation<Table>()!!
+
+    val accessAnnotationEntry =
+        accessAnnotationEntries.find { it.annotation == accessAnnotationClass }
+    val allowedCapabilities = accessAnnotationEntry?.capabilities?.toSet() ?: setOf()
+    val allowedServices = accessAnnotationEntry?.services?.toSet() ?: setOf()
+
+    val constraintsResult = mutableListOf<Pair<Type, DatabaseQueryMetadata.ConstraintMetadata>>()
```

Comment:
Should `Type` be a field of `ConstraintMetadata` ?

Reply:
I'd lean against since `ConstraintMetadata` and `Type` support two very different parts of the front end and smoshing them together would lead to a bit more confusion and transforming in the web code.